pkg/block/indexheader/binary_reader.go

```diff
@@ -476,7 +464,11 @@ func (w *binaryWriter) Write(p []byte) (int, error) {
 }
 
 func (w *binaryWriter) Buffer() []byte {
-	return w.writer.Buffer()
+	pwb, ok := w.writer.(PosWriterWithBuffer)
```

Comment:
This looks a bit like a hack. Where is the `Buffer` method called? Maybe we can just use a `PosWriterWithBuffer` there instead.

Reply:
It is used here https://github.com/thanos-io/thanos/blob/main/pkg/block/indexheader/binary_reader.go#L130.

It will only be called if it is a `MemWriter`. If it is file writer mode it returns previously. I think the design is not the best though and needs more refactoring.

If we think this kind of hack is not ok, I can add the `Buffer` method back to `FileWriter` and just return nil

Reply:
Is it possible to write the bytes to both the file writer and the memory writer? So we would need some sort of a composite `Writer` that writes to multiple `Writer`s.

Reply:
> Is it possible to write the bytes to both the file writer and the memory writer?

It is the current implementation and I don't think it is correct. We don't have to write memory writer if we have the file writer IIUC. It is just holding the extra memory for nothing. (We are reading from the file anyway)

Reply:
Hm then why do we need the buffer? Does it hold the meta file temporarily?

Reply:
The `buffer` is used by memory writer only because it holds the index header all in memory. 
It is not used in file writer at all. But the current interface `PosWriter` has `buffer` method so both file writer and memory writer implements it.

This pr basically removes that buffer from file writer. But I don't have a good way to refactor the `PosWriter` interface to remove `buffer` method so I added `PosWriterWithBuffer`.

Reply:
https://github.com/thanos-io/thanos/blob/main/pkg/block/indexheader/binary_reader.go#L547 As you can see, the first returned value is the buffer but it is ignored if it is file mode. Otherwise, at line 554 the returned buffer will be used.

Reply:
@fpetkovski Does it make sense overall? TLDR is that buffer is not used at all for file writer.